src/typechecker.erl

```diff
@@ -2012,20 +2012,20 @@ do_type_check_expr_in(Env, ResTy, {tuple, LINE, TS}) ->
                                            || Expr <- TS ]),
             {union_var_binds(VBs, Env#env.tenv), constraints:combine(Css)};
         {type_error, _} ->
-            throw({type_error, tuple, LINE, ResTy})
+            throw({type_error, Tuple, type(tuple, any), ResTy})
     end;
 
 %% Maps
-do_type_check_expr_in(Env, ResTy, {map, LINE, Assocs}) ->
-    case subtype(ResTy, {type, LINE, map, any}, Env#env.tenv) of
+do_type_check_expr_in(Env, ResTy, {map, _, Assocs} = Map) ->
+    case subtype(ResTy, type(map, any), Env#env.tenv) of
```

Comment:
this was wrong in the original code: should be `map() :: ResTy` instead of `ResTy :: map()` (I think `record_field` and `record_index` has similar flip issue)

Reply:
I don't quite follow, could you spell out what the code should say?

Reply:
What I meant is `subtype(ResTy, type(map, any),...` checks whether `ResTy` is a subtype of `map()`, but we should rather check that `map()` is a subtype of `ResTy` (ie `subtype(, type(map, any), ResTy,...`)

Reply:
I'm afraid neither is correct. Consider
```erlang
-spec f() -> #{ a := b } | undefined.
f() -> #{ a => b }
```
We have neither `map() :: #{a := b} | undefined` or the other way around. Instead we need a `expect_map_type` or similar (perhaps taking the `glb` of `map()` and `ResTy` first?) to dig out the expected types for the assocs. But this isn't really the PR for that... I've filed #131.

Reply:
Hey, we do have `map() :: #{a := b} | undefined`. Of course it's not complete as we don't check the keys and values, but `map()` is the any-map so it's a compatible subtype to `#{a := b}`.

Reply:
Oh, you're right. I still get an error on the example though... mysterious.